// isotope-client/login/LoginForm.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactRouterRedux = require('react-router-redux');

var _FloatingActionButton = require('material-ui/FloatingActionButton');

var _FloatingActionButton2 = _interopRequireDefault(_FloatingActionButton);

var _arrowForward = require('material-ui/svg-icons/navigation/arrow-forward');

var _arrowForward2 = _interopRequireDefault(_arrowForward);

var _reduxForm = require('redux-form');

var _reduxFormMaterialUi = require('redux-form-material-ui');

var _loginActions = require('./loginActions');

var _loginApi = require('./loginApi');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

// eslint-disable-line max-len

var style = {
	button: {
		position: 'absolute',
		top: '50%',
		right: '-26px',
		marginTop: '-26px'
	}
};

var validate = function validate(values) {
	var errors = {};
	if (!values.login) {
		errors.login = "L'identifiant est obligatoire";
	}
	if (!values.password) {
		errors.password = 'Le mot de passe est obligatoire';
	}
	return errors;
};

var LoginForm = function (_Component) {
	_inherits(LoginForm, _Component);

	function LoginForm(props) {
		_classCallCheck(this, LoginForm);

		var _this = _possibleConstructorReturn(this, (LoginForm.__proto__ || Object.getPrototypeOf(LoginForm)).call(this, props));

		_this.loginSubmit = _this.loginSubmit.bind(_this);
		return _this;
	}

	_createClass(LoginForm, [{
		key: 'loginSubmit',
		value: function loginSubmit(data, dispatch) {
			var nextPathname = this.props.nextPathname;

			return new Promise(function (resolve, reject) {
				return (0, _loginApi.login)(data).then(function (json) {
					if (json) {
						dispatch((0, _loginActions.logUser)(json.token));
						(0, _loginApi.storeLocalUser)(json);
						resolve(json.token);
					} else {
						reject({
							login: 'Mauvais couple identifiant / mot de passe'
						});
					}
				}).catch(function () {
					reject({
						login: 'Connexion impossible'
					});
				});
			}).then(function (token) {
				return (0, _loginApi.fetchUser)(token);
			}).then(function (user) {
				dispatch((0, _loginActions.getUser)(user));
			}).then(function () {
				if (nextPathname) {
					dispatch((0, _reactRouterRedux.push)(nextPathname));
				} else {
					dispatch((0, _reactRouterRedux.push)('/'));
				}
			});
		}
	}, {
		key: 'render',
		value: function render() {
			var handleSubmit = this.props.handleSubmit;

			return _react2.default.createElement(
				'form',
				{ onSubmit: handleSubmit(this.loginSubmit), autoComplete: 'off' },
				_react2.default.createElement(_reduxForm.Field, {
					component: _reduxFormMaterialUi.TextField,
					floatingLabelText: 'Identifiant',
					hintText: 'Entrer votre identifiant',
					name: 'login',
					fullWidth: true,
					defaultValue: undefined
				}),
				_react2.default.createElement(_reduxForm.Field, {
					component: _reduxFormMaterialUi.TextField,
					type: 'password',
					hintText: 'Entrer votre mot de passe',
					floatingLabelText: 'Mot de passe',
					name: 'password',
					fullWidth: true,
					defaultValue: undefined
				}),
				_react2.default.createElement(
					_FloatingActionButton2.default,
					{ style: style.button, type: 'submit' },
					_react2.default.createElement(_arrowForward2.default, null)
				)
			);
		}
	}]);

	return LoginForm;
}(_react.Component);

LoginForm.propTypes = {
	handleSubmit: _react.PropTypes.func.isRequired,
	nextPathname: _react.PropTypes.string
};

var mapStateToProps = function mapStateToProps(state) {
	return { user: state.user };
};

exports.default = (0, _reduxForm.reduxForm)({
	form: 'loginForm',
	validate: validate
}, mapStateToProps)(LoginForm);
