// isotope-client/users/table/UsersTable.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _IsotopeTable = require('../../components/table/IsotopeTable');

var _IsotopeTable2 = _interopRequireDefault(_IsotopeTable);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var UsersTable = function UsersTable(_ref) {
	var users = _ref.users,
	    editAction = _ref.editAction,
	    deleteAction = _ref.deleteAction;
	return _react2.default.createElement(_IsotopeTable2.default, {
		liste: users,
		colonnes: [{ value: 'login', label: 'Identifiant' }, { value: 'lastname', label: 'Nom' }, { value: 'firstname', label: 'Prénom' }, { value: 'email', label: 'Mail' }, { value: 'phoneNumber', label: 'Téléphone' }, { value: 'company', label: 'Société' }, { value: 'role', label: 'Profil' }],
		actions: [{ label: 'Modifier', action: function action(user) {
				return editAction(user.id);
			} }, { label: 'Supprimer', action: function action(user) {
				return deleteAction(user.id);
			} }]
	});
};

UsersTable.propTypes = {
	users: _react.PropTypes.array,
	editAction: _react.PropTypes.func,
	deleteAction: _react.PropTypes.func
};

exports.default = UsersTable;
