// isotope-client/components/table/IsotopeTable.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Table = require('material-ui/Table');

var _IsotopeTableMenu = require('./IsotopeTableMenu');

var _IsotopeTableMenu2 = _interopRequireDefault(_IsotopeTableMenu);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var IsotopeTable = function IsotopeTable(_ref) {
	var liste = _ref.liste,
	    colonnes = _ref.colonnes,
	    actions = _ref.actions,
	    key = _ref.key;
	return _react2.default.createElement(
		_Table.Table,
		null,
		_react2.default.createElement(
			_Table.TableHeader,
			{
				displaySelectAll: false,
				adjustForCheckbox: false
			},
			_react2.default.createElement(
				_Table.TableRow,
				null,
				(colonnes || []).map(function (colonne, index) {
					return _react2.default.createElement(
						_Table.TableHeaderColumn,
						{ key: index },
						colonne.label
					);
				}),
				actions && _react2.default.createElement(_Table.TableHeaderColumn, null)
			)
		),
		_react2.default.createElement(
			_Table.TableBody,
			{
				displayRowCheckbox: false
			},
			(liste || []).map(function (element) {
				return _react2.default.createElement(
					_Table.TableRow,
					{ key: key ? element[key] : element.id },
					(colonnes || []).map(function (colonne, index) {
						return _react2.default.createElement(
							_Table.TableRowColumn,
							{ key: index },
							colonne.component || element[colonne.value]
						);
					}),
					actions && _react2.default.createElement(
						_Table.TableRowColumn,
						null,
						_react2.default.createElement(_IsotopeTableMenu2.default, { actions: actions, element: element })
					)
				);
			})
		)
	);
};

IsotopeTable.propTypes = {
	/**
  * Liste des éléments à afficher
  */
	liste: _react.PropTypes.array.isRequired,
	/**
  * Clé des éléments à utiliser pour l'identification sur les lignes
  */
	key: _react.PropTypes.string,
	/**
  * Objet JSON qui décrit les colonnes à utiliser avec la table sous la forme `{
  *  value: propriété à utiliser dans l'élément en cours (objet de la liste),
  *  label: entête de la colonne utilisé,
  *  component: //Optionnel : permet d'utiliser un composant dans le TD
  * }`
  *
  */
	colonnes: _react.PropTypes.array.isRequired,
	/**
  * Liste des actions afficher en fin de ligne. Une action est un objet JSON de la forme<br/>
  * `{
  *   label: '',
  *   action: (element) => console.log('')
  * }`
  */
	actions: _react.PropTypes.array
};

exports.default = IsotopeTable;
