// isotope-client/components/table/IsotopeTableMenu.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _IconMenu = require('material-ui/IconMenu');

var _IconMenu2 = _interopRequireDefault(_IconMenu);

var _MenuItem = require('material-ui/MenuItem');

var _MenuItem2 = _interopRequireDefault(_MenuItem);

var _IconButton = require('material-ui/IconButton/IconButton');

var _IconButton2 = _interopRequireDefault(_IconButton);

var _moreVert = require('material-ui/svg-icons/navigation/more-vert');

var _moreVert2 = _interopRequireDefault(_moreVert);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Components
var IsotopeTableMenu = function IsotopeTableMenu(_ref) {
	var actions = _ref.actions,
	    element = _ref.element;
	return _react2.default.createElement(
		_IconMenu2.default,
		{
			iconButtonElement: _react2.default.createElement(
				_IconButton2.default,
				null,
				_react2.default.createElement(_moreVert2.default, null)
			),
			anchorOrigin: { horizontal: 'right', vertical: 'top' },
			targetOrigin: { horizontal: 'left', vertical: 'top' }
		},
		(actions || []).map(function (action, index) {
			return _react2.default.createElement(_MenuItem2.default, { key: index, primaryText: action.label, onTouchTap: function onTouchTap() {
					return action.action(element);
				} });
		})
	);
};

IsotopeTableMenu.propTypes = {
	actions: _react.PropTypes.array,
	element: _react.PropTypes.object.isRequired
};

exports.default = IsotopeTableMenu;
