// isotope-client/users/UsersManagement.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactRedux = require('react-redux');

var _add = require('material-ui/svg-icons/content/add');

var _add2 = _interopRequireDefault(_add);

var _usersActions = require('./usersActions');

var actions = _interopRequireWildcard(_usersActions);

var _usersSelectors = require('./usersSelectors');

var selectors = _interopRequireWildcard(_usersSelectors);

var _UsersTable = require('./table/UsersTable');

var _UsersTable2 = _interopRequireDefault(_UsersTable);

var _UserForm = require('./form/UserForm');

var _UserForm2 = _interopRequireDefault(_UserForm);

var _RightBottomFixedButton = require('../components/RightBottomFixedButton');

var _ConfirmationSuppression = require('../components/ConfirmationSuppression');

var _ConfirmationSuppression2 = _interopRequireDefault(_ConfirmationSuppression);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var UsersManagement = function (_Component) {
	_inherits(UsersManagement, _Component);

	function UsersManagement() {
		_classCallCheck(this, UsersManagement);

		return _possibleConstructorReturn(this, (UsersManagement.__proto__ || Object.getPrototypeOf(UsersManagement)).apply(this, arguments));
	}

	_createClass(UsersManagement, [{
		key: 'componentWillMount',
		value: function componentWillMount() {
			this.props.getUsers();
		}
	}, {
		key: 'render',
		value: function render() {
			var _props = this.props,
			    users = _props.users,
			    user = _props.user,
			    editUser = _props.editUser,
			    deleteUser = _props.deleteUser,
			    addUser = _props.addUser,
			    idDeletingUser = _props.idDeletingUser,
			    cancelDeleteUser = _props.cancelDeleteUser,
			    saveDeleteUser = _props.saveDeleteUser;


			return _react2.default.createElement(
				'div',
				null,
				_react2.default.createElement(
					_RightBottomFixedButton.RightBottomFixedButton,
					{ label: 'Ajouter', onTouchTap: addUser },
					_react2.default.createElement(_add2.default, null)
				),
				_react2.default.createElement(_UsersTable2.default, { users: users, editAction: editUser, deleteAction: deleteUser }),
				user && _react2.default.createElement(_UserForm2.default, null),
				idDeletingUser > 0 && _react2.default.createElement(_ConfirmationSuppression2.default, {
					title: 'Suppression de l\'utilisateur',
					annulerSuppression: cancelDeleteUser,
					suppression: function suppression() {
						return saveDeleteUser(idDeletingUser);
					}
				})
			);
		}
	}]);

	return UsersManagement;
}(_react.Component);

UsersManagement.propTypes = {
	getUsers: _react.PropTypes.func.isRequired,
	editUser: _react.PropTypes.func,
	deleteUser: _react.PropTypes.func,
	cancelDeleteUser: _react.PropTypes.func,
	saveDeleteUser: _react.PropTypes.func,
	addUser: _react.PropTypes.func,
	users: _react.PropTypes.array,
	user: _react.PropTypes.object,
	idDeletingUser: _react.PropTypes.number
};

UsersManagement.contextTypes = {
	muiTheme: _react.PropTypes.object.isRequired
};

var mapStateToProps = function mapStateToProps(state) {
	return {
		users: selectors.getUsers(state),
		user: selectors.getEditingUser(state),
		idDeletingUser: selectors.getIdDeletingUser(state)
	};
};

var dispatchToProps = {
	getUsers: actions.getUsers,
	editUser: actions.editUser,
	addUser: actions.addUser,
	deleteUser: actions.deleteUser,
	cancelDeleteUser: actions.cancelDeleteUser,
	saveDeleteUser: actions.saveDeleteUser
};

exports.default = (0, _reactRedux.connect)(mapStateToProps, dispatchToProps)(UsersManagement);
