// isotope-client/components/menu/menuActions.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.getMenu = exports.GET_MENU_ERROR = exports.GET_MENU_SUCCESS = exports.GET_MENU_REQUEST = undefined;

var _menuApi = require('./menuApi');

var api = _interopRequireWildcard(_menuApi);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

var GET_MENU_REQUEST = exports.GET_MENU_REQUEST = 'GET_MENU_REQUEST';
var GET_MENU_SUCCESS = exports.GET_MENU_SUCCESS = 'GET_MENU_SUCCESS';
var GET_MENU_ERROR = exports.GET_MENU_ERROR = 'GET_MENU_ERROR';
var getMenu = exports.getMenu = function getMenu(code) {
	return function (dispatch) {
		dispatch({ type: GET_MENU_REQUEST });
		return api.getMenu(code).then(function (menu) {
			return dispatch({
				type: GET_MENU_SUCCESS,
				menu: { code: code, entries: menu }
			});
		}).catch(function (e) {
			return dispatch({ type: GET_MENU_ERROR });
		});
	};
};
