// isotope-client/components/AppBar.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _AppBar = require('material-ui/AppBar');

var _AppBar2 = _interopRequireDefault(_AppBar);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var getStyles = function getStyles(props, context) {
	var palette = context.muiTheme.palette;

	return {
		bar: {
			position: 'fixed',
			background: palette.accent1Color,
			top: 0
		}
	};
};

var AppBar = function AppBar() {
	var props = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
	var context = arguments[1];

	var title = props.title,
	    style = props.style,
	    otherProps = _objectWithoutProperties(props, ['title', 'style']);

	var styles = getStyles(props, context);
	return _react2.default.createElement(
		'div',
		null,
		_react2.default.createElement(_AppBar2.default, _extends({
			title: title,
			zDepth: 1,
			style: _extends({}, styles.bar, style),
			showMenuIconButton: false
		}, otherProps))
	);
};

AppBar.propTypes = {
	title: _react.PropTypes.string.isRequired,
	style: _react.PropTypes.object
};

AppBar.contextTypes = {
	muiTheme: _react.PropTypes.object.isRequired
};

exports.default = AppBar;
