// isotope-client/components/ConfirmationSuppression.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Dialog = require('material-ui/Dialog');

var _Dialog2 = _interopRequireDefault(_Dialog);

var _FlatButton = require('material-ui/FlatButton');

var _FlatButton2 = _interopRequireDefault(_FlatButton);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ConfirmationSuppression = function ConfirmationSuppression(_ref) {
	var children = _ref.children,
	    annulerSuppression = _ref.annulerSuppression,
	    suppression = _ref.suppression,
	    title = _ref.title;
	return _react2.default.createElement(
		_Dialog2.default,
		{
			actions: [_react2.default.createElement(_FlatButton2.default, {
				label: 'Annuler',
				primary: true,
				onTouchTap: annulerSuppression
			}), _react2.default.createElement(_FlatButton2.default, {
				label: 'Valider',
				primary: true,
				onTouchTap: suppression
			})],
			open: true,
			modal: true
		},
		title
	);
};

ConfirmationSuppression.propTypes = {
	suppression: _react.PropTypes.func.isRequired,
	annulerSuppression: _react.PropTypes.func.isRequired,
	title: _react.PropTypes.string.isRequired
};

exports.default = ConfirmationSuppression;
