// isotope-client/components/EmptyPage.js
"use strict";

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _localFlorist = require("material-ui/svg-icons/maps/local-florist");

var _localFlorist2 = _interopRequireDefault(_localFlorist);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var EmptyPage = function EmptyPage(_ref, _ref2) {
	var title = _ref.title,
	    _ref$style = _ref.style,
	    style = _ref$style === undefined ? {} : _ref$style;
	var muiTheme = _ref2.muiTheme;

	var innerStyle = {
		container: {
			position: 'absolute',
			left: 0,
			right: 0,
			top: 0,
			bottom: 0,
			textAlign: 'center',
			display: 'flex',
			alignItems: 'center',
			zIndex: 1
		},
		content: {
			flexBasis: '50%',
			flex: 1
		},
		p: {
			color: muiTheme.palette.disabledColor
		}
	};

	var styleContainer = _extends({}, innerStyle.container, style);
	return _react2.default.createElement(
		"div",
		{ style: styleContainer },
		_react2.default.createElement(
			"div",
			{ style: innerStyle.content },
			_react2.default.createElement(_localFlorist2.default, { color: muiTheme.palette.disabledColor, style: { width: 120, height: 120 } }),
			_react2.default.createElement(
				"p",
				{ style: innerStyle.p },
				title
			)
		)
	);
};

EmptyPage.propTypes = {
	title: _react.PropTypes.string.isRequired,
	style: _react.PropTypes.object
};
EmptyPage.contextTypes = {
	muiTheme: _react.PropTypes.object.isRequired
};

exports.default = EmptyPage;
