// isotope-client/components/download/IsotopeLink.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.download = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _fetchEntities = require('../../entities/fetchEntities');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var download = exports.download = function download(href) {
	(0, _fetchEntities.fetchFactory)('/ticket', {
		method: 'POST',
		body: JSON.stringify({ path: href })
	}, '/api').then(function (ticket) {
		if (href.indexOf('?') == -1) {
			window.location = href + '?ticket=' + ticket.ticket;
		} else {
			window.location = href + '&ticket=' + ticket.ticket;
		}
	});
};

var isotopeLink = function isotopeLink(_ref) {
	var children = _ref.children,
	    href = _ref.href;
	return _react2.default.createElement(
		'span',
		{ onTouchTap: function onTouchTap() {
				return download(href);
			} },
		children
	);
};

isotopeLink.propTypes = {
	/**
  * Contenu du link
  */
	children: _react.PropTypes.node.isRequired,
	/**
  * Lien vers le document à télécharger
  */
	href: _react.PropTypes.string.isRequired
};

exports.default = isotopeLink;
